### PR TITLE
[SPARK-12507][Streaming][Document]Expose closeFileAfterWrite and  allowBatching configurations for Streaming

### PR DESCRIPTION
**Pull request by zsxwing**
/cc @tdas @brkyvz

### CODE REVIEW

**Comment by srowen** [comment]
Let's improve the title of items like this. "Update x" is never descriptive

File: docs/configuration.md [hunk 1]
```diff
@@ -1600,6 +1600,33 @@ Apart from these, the following properties are also available, and may be useful
     How many batches the Spark Streaming UI and status APIs remember before garbage collecting.
   </td>
 </tr>
+<tr>
+  <td><code>spark.streaming.driver.writeAheadLog.closeFileAfterWrite</code></td>
+  <td>false</td>
+  <td>
+    Whether to close the file after writing a write ahead log record in driver. Because S3 doesn't
```
Comment (BenFradet) [inline docs/configuration.md]: I'd say `on the driver` instead of `in driver`.

File: docs/configuration.md [hunk 2]
```diff
@@ -1600,6 +1600,33 @@ Apart from these, the following properties are also available, and may be useful
...
+    support flushing of data, when using S3 for checkpointing, you should enable it to achieve read
+    after write consistency.
+  </td>
+</tr>
+<tr>
+  <td><code>spark.streaming.receiver.writeAheadLog.closeFileAfterWrite</code></td>
+  <td>false</td>
+  <td>
+    Whether to close the file after writing a write ahead log record in receivers. Because S3
```
Comment (BenFradet) [inline docs/configuration.md]: same thing here: `on the receivers` instead of `in receivers`

File: docs/configuration.md [hunk 3]
```diff
@@ -1600,6 +1600,33 @@ Apart from these, the following properties are also available, and may be useful
...
+    doesn't support flushing of data, when using S3 for checkpointing, you should enable it to
+    achieve read after write consistency.
+  </td>
+</tr>
+<tr>
+  <td><code>spark.streaming.driver.writeAheadLog.allowBatching</code></td>
+  <td>false</td>
+  <td>
+    Whether to batch write ahead logs in driver to write. When using S3 for checkpointing, write
```
Comment (BenFradet) [inline docs/configuration.md]: Here, I'd say `on the driver` instead of `in driver to write`.

File: docs/configuration.md [hunk 4]
```diff
@@ -1600,6 +1600,33 @@ Apart from these, the following properties are also available, and may be useful
...
+    operations in driver usually take too long. Enable batching write ahead logs will improve
```
Comment (BenFradet) [inline docs/configuration.md]: same: `on the`
and `Enabling` instead of `Enable`

File: docs/configuration.md [hunk 5]
```diff
@@ -1600,6 +1600,33 @@ Apart from these, the following properties are also available, and may be useful
...
+    the performance of writing.
```
Comment (BenFradet) [inline docs/configuration.md]: I'd say `will improve the performance of write operations`

File: docs/streaming-programming-guide.md [hunk 6]
```diff
@@ -2029,6 +2029,11 @@ If the data is being received by the receivers faster than what can be processed
 you can limit the rate by setting the [configuration parameter](configuration.html#spark-streaming)
 `spark.streaming.receiver.maxRate`.
 
+If using S3 for checkpointing, please remember to enable `spark.streaming.driver.writeAheadLog.closeFileAfterWrite`
+and `spark.streaming.receiver.writeAheadLog.closeFileAfterWrite`. You can also enable
+`spark.streaming.driver.writeAheadLog.allowBatching` to improve the performance of writing write
+ahead logs in driver. See [Spark Streaming Configuration](configuration.html#spark-streaming) or more details.
```
Comment (BenFradet) [inline docs/streaming-programming-guide.md]: `on the driver` and `for more details`

**Comment by BenFradet** [comment]
I have a few comments on phrasing but otherwise it lgtm

File: docs/configuration.md [hunk 7]
```diff
@@ -1600,6 +1600,33 @@ Apart from these, the following properties are also available, and may be useful
...
+  <td>false</td>
+  <td>
+    Whether to close the file after writing a write ahead log record in receivers. Because S3
+    doesn't support flushing of data, when using S3 for checkpointing, you should enable it to
+    achieve read after write consistency.
+  </td>
+</tr>
+<tr>
+  <td><code>spark.streaming.driver.writeAheadLog.allowBatching</code></td>
+  <td>false</td>
```
Comment (zsxwing) [inline docs/configuration.md]: for me: the default value is `true`.

That's why I want to expose this one since the behavior is different from 1.5.0.

**Comment by zsxwing** [comment]
@BenFradet Addressed. Thanks for your reviewing.

**Comment by brkyvz** [comment]
LGTM

**Comment by brkyvz** [comment]
Maybe we can also include that `allowBatching` is not just helpful when `closeFileAfterWrite` is enabled, but is also very helpful to scale to a large number of receivers 50+ for example.

**Comment by zsxwing** [comment]
> Maybe we can also include that allowBatching is not just helpful when closeFileAfterWrite is enabled, but is also very helpful to scale to a large number of receivers 50+ for example.

I guess `enabled` should be `disabled`?

**Comment by brkyvz** [comment]
@zsxwing Thanks! LGTM

File: docs/configuration.md [hunk 8]
```diff
@@ -1600,6 +1600,34 @@ Apart from these, the following properties are also available, and may be useful
...
+    Whether to close the file after writing a write ahead log record on the driver. Because S3 doesn't
+    support flushing of data, when using S3 for checkpointing, you should enable it to achieve read
+    after write consistency.
+  </td>
+</tr>
+<tr>
+  <td><code>spark.streaming.receiver.writeAheadLog.closeFileAfterWrite</code></td>
+  <td>false</td>
+  <td>
+    Whether to close the file after writing a write ahead log record on the receivers. Because S3
```
Comment (tdas) [inline docs/configuration.md]: Because file systems like S3

File: docs/configuration.md [hunk 9]
```diff
@@ -1600,6 +1600,34 @@ Apart from these, the following properties are also available, and may be useful
     How many batches the Spark Streaming UI and status APIs remember before garbage collecting.
   </td>
 </tr>
+<tr>
+  <td><code>spark.streaming.driver.writeAheadLog.closeFileAfterWrite</code></td>
+  <td>false</td>
+  <td>
+    Whether to close the file after writing a write ahead log record on the driver. Because S3 doesn't
```
Comment (tdas) [inline docs/configuration.md]: Rewrite similar to below, replacing driver with receiver.

File: docs/streaming-programming-guide.md [hunk 10]
```diff
@@ -2029,6 +2029,11 @@ If the data is being received by the receivers faster than what can be processed
 you can limit the rate by setting the [configuration parameter](configuration.html#spark-streaming)
 `spark.streaming.receiver.maxRate`.
 
+If using S3 for checkpointing, please remember to enable `spark.streaming.driver.writeAheadLog.closeFileAfterWrite`
```
Comment (tdas) [inline docs/streaming-programming-guide.md]: "If" --> "while"
checkpointing --> Write Ahead Logs

File: docs/streaming-programming-guide.md [hunk 11]
```diff
@@ -2029,6 +2029,11 @@ If the data is being received by the receivers faster than what can be processed
 you can limit the rate by setting the [configuration parameter](configuration.html#spark-streaming)
```
Comment (tdas) [inline docs/streaming-programming-guide.md]: Can you remove this section completely? 
- Remove the rate stuff as it has already been covered earlier. 
- Put the closeAfterFileWrite stuff in the WAL sections above. Keeps things in context.

File: docs/configuration.md [hunk 12]
```diff
@@ -1600,6 +1600,34 @@ Apart from these, the following properties are also available, and may be useful
...
+    doesn't support flushing of data, when using S3 for checkpointing, you should enable it to
+    achieve read after write consistency.
+  </td>
+</tr>
+<tr>
+  <td><code>spark.streaming.driver.writeAheadLog.allowBatching</code></td>
+  <td>true</td>
+  <td>
+    Whether to batch write ahead logs on the driver to write. When using S3 for checkpointing, write
+    operations on the driver usually take too long. Enabling batching write ahead logs will improve
```
Comment (tdas) [inline docs/configuration.md]: Whether to batch writes to the metadata WAL in the driver. This is useful from improving performance for file systems like S3 where the write latency is high and / or scenarios with large number of receivers.

File: docs/streaming-programming-guide.md [hunk 13]
```diff
@@ -1985,7 +1985,11 @@ To run a Spark Streaming applications, you need to have the following.
   to increase aggregate throughput. Additionally, it is recommended that the replication of the
   received data within Spark be disabled when the write ahead log is enabled as the log is already
   stored in a replicated storage system. This can be done by setting the storage level for the
-  input stream to `StorageLevel.MEMORY_AND_DISK_SER`.
+  input stream to `StorageLevel.MEMORY_AND_DISK_SER`. While using S3 (or any file system that
+  does not support flushing) for Write Ahead Logs, please remember to enable
```
Comment (tdas) [inline docs/streaming-programming-guide.md]: nit: Write Ahead Logs is not in caps in this text. so please be consistent.

**Comment by tdas** [comment]
just one more comment. then LGTM.

**Comment by tdas** [comment]
LGTM. Merging this to master and 1.6. Thanks!

**Pull request closed without merging**